### PR TITLE
Implement the Date standard library

### PR DESCRIPTION
**Pull request by LiaoPeng**
Implement the date standard library

Finished:
- [x]  getFullYear, getMonth, getDate, getDay, getHours, getMinutes, getSeconds, getMilliseconds
- [x]  setFullYear, setMonth, setDate, setHours, setMinutes, setSeconds, setMilliseconds
- [x]  get UTC timestamp
- [x] enhancement the effective
- [x] getTimezoneOffset

### CODE REVIEW

File: std/assembly/internal/date.ts [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+  static floorDivide(n: i32, d: i32): i32 {
+    return ((n >= 0) ? (n / d) : (((n + 1) / d) - 1));
+  }
+
+  static mod(x: i32, y: i32): i32 {
+    return (x - y * CalendarUtils.floorDivide(x, y));
+  }
+
+  static isLeapYear(gregorianYear: i32): bool {
+    return  (((gregorianYear % 4) == 0)
```
Comment (MaxGraey) [inline std/assembly/internal/date.ts]: I think also need add this check before other `gregorianYear > 0` && ...

File: std/assembly/internal/date.ts [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+       * @param dayOfMonth the day of the month, 1-based
+       */
+      private getFixedDate(year: i32, month: i32, dayOfMonth: i32): i32 {
+
+        var prevyear = year - 1;
+        var days = dayOfMonth;
+
+        if (prevyear >= 0) {
+          days += (365 * prevyear)
+          + (prevyear / 4)
```
Comment (MaxGraey) [inline std/assembly/internal/date.ts]: How about create separate util function like:
```ts
@inline
function countLeaps(year: i32): i32 {
  --year;
  return year / 4 - year / 100 + year / 400;
}
```
And now calc days for example will be as:
```ts
days = (year - 1) * 365 + countLeaps(year) + ((367 * month - 362) / 12);
```

File: std/assembly/internal/date.ts (hunk 1 above)
Comment (MaxGraey) [reply std/assembly/internal/date.ts]: What about speed up this little bit?
```ts
static isLeapYear(year: i32): bool {
  return (year & 3) == 0 && (((year % 25) | 0) != 0 || (year & 15) == 0);
}
```

Also not sure about additional check  `gregorianYear > 0`. Need more investigate

File: std/assembly/internal/date.ts [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+          - (prevyear / 100)
+          + (prevyear / 400)
+          + ((367 * month - 362) / 12);
```
Comment (MaxGraey) [inline std/assembly/internal/date.ts]: Hmm, btw `moment` it seems calculate this as `365 * year + countLeaps(year)`:
http://jsfiddle.net/qb3d9pzs

File: std/assembly/internal/date.ts (hunk 1 above)
Comment (LiaoPeng) [reply std/assembly/internal/date.ts]: The implement to judge the leap year is cool.

File: std/assembly/internal/date.ts (hunk 2 above)
Comment (LiaoPeng) [reply std/assembly/internal/date.ts]: Good suggestion.

File: std/assembly/date.ts [hunk 4]
```diff
@@ -22,17 +25,111 @@ export class Date {
...
   setTime(value: i64): i64 {
-    this.value = value;
+    if (this.value != value) {
+      this.value = value;
+      this.normolize();
+    }
     return value;
   }
+
+  private normolize(): void {
```
Comment (dcodeIO) [inline std/assembly/date.ts]: Nit: normalize

**Comment by LiaoPeng** [comment]
It's not necessary. I would remove the `BaseCalendar`.

**Comment by LiaoPeng** [comment]
I have a question how to get the local TimeZone.  It blocked me.

**Comment by MaxGraey** [comment]
Timezone offset is just difference between `Date.UTC` and `Date.now`

**Comment by LiaoPeng** [comment]
@MaxGraey 
The method `getTimezoneOffset` Return the timezone difference between UTC and Local Time.
Both `Date.utc` and `Date.now` return the UTC time.  I am not sure that's ok.

**Comment by dcodeIO** [comment]
It's a shame that I haven't reviewed this more thoroughly. Looking at it again I think that this went stale in my head because it needs so many managed objects to represent a single date, and a more C-like implementation with data living on the date objects felt more efficient. In general I think that this is a good approach as it passes all the tests, so wondering if you are still interested in continuing on this PR?

**Comment by dcodeIO** [comment]
That would be fantastic, yeah. To elaborate a bit on the C-likeness, there currently are instances of `CalendarDate` and `TimeZone` required to represent a `Date`, so there's quite a bit of managed objects to reference count and collect. Ideally I think that having a `std/util/calendar.ts` exposing plain functions to work with let's say `i64` values representing these would be more in-line with other stdlib internals, so that we have just a single allocation per `Date` instance to keep track of. Do you think it'd be possible to transform the existing code into this direction?

**Comment by MaxGraey** [comment]
Also it will be ideal if you organize the code in such a way that it could be reused for Temporal proposal (https://github.com/tc39/proposal-temporal/tree/main/polyfill/lib)

**Comment by dcodeIO** [comment]
Oh, wasn't aware of this one. Interesting.

**Comment by torch2424** [comment]
@LiaoPeng Ah! Thank you very much for opening this! :smile: This is a ton of awesome work, it is super appreciated! :smile: 

While you work with @dcodeIO and @MaxGraey on building this out as a c-style API, and if you think it would take a while. I'd be interested in pulling this Date implementation into it's own library that could be `npm install`'d? I know it's not the most optimal for performance currently, but I think it'd be useful to a few folks in the communitty in the meantime :smile: 

That is of course, if you wouldn't mind me doing so (giving this PR and you full credit @LiaoPeng ) ? Or we could start it together? Whichever works! 

Thanks y'all! :smile:

**Comment by LiaoPeng** [comment]
@torch2424 Hey，it would take a while. Your advice is good, just do it.  Now I push new commit `rebased` the current master branch.

**Comment by torch2424** [comment]
@LiaoPeng Ah! Thank you for rebasing! :smile: :pray: 

> Your advice is good, just do it.

So just to confirm, you're okay with me using the code in the PR on a `as-date` repo? (Pointing back to you and this PR of course) :smile:

**Comment by LiaoPeng** [comment]
@torch2424 I am okey,

**Comment by dcodeIO** [comment]
Various Date functions have been implemented meanwhile, avoiding intermediary allocations.

**Pull request closed without merging**